tsinfer/inference.py

```diff
@@ -490,6 +607,111 @@ def match_samples(
+    if sample_data.sequence_length != tables.sequence_length:
+        raise ValueError(
+            f"The sample_data and tables must have the same sequence length"
+        )
+    ts = tables.tree_sequence()
+    if check_consistency:
+        verify_samples(sample_data, ts, sample_id_map=sample_id_map)
+    trees = ts.trees()
+    tree = next(trees)
+    sample_indexes = np.array(list(sample_id_map.keys()), dtype=int)
+    positions = sample_data.sites_position[:]
+    new_sd_sites = np.where(np.isin(positions, tables.sites.position) == 0)[0]
+    # Create new sites and add the mutations
+    pm = progress_monitor.get("ms_parsimony_mutations", len(new_sd_sites))
+    for variant in sample_data.variants(ids=new_sd_sites):
+        site = variant.site
+        pos = site.position
+        anc_state = site.ancestral_state
+        anc_value = 0  # sample_data files always have 0 as the ancestral allele idx
+        G = variant.genotypes[sample_indexes]
+        if np.all(np.logical_or(G == tskit.MISSING_DATA, G == anc_value)):
+            tables.sites.add_row(
+                position=pos,
+                ancestral_state=anc_state,
+                metadata=_encode_metadata(
+                    _update_site_metadata(
+                        site.metadata, inference_type=constants.INFERENCE_NONE
+                    )
+                ),
+            )
+        else:
+            while tree.interval[1] <= pos:
+                tree = next(trees)
+            inferred_anc_state, mapped_mutations = tree.map_mutations(
+                G, variant.alleles
+            )
+            if anc_state is None:
+                anc_state = inferred_anc_state
+            new_site_id = tables.sites.add_row(
+                position=pos,
+                ancestral_state=anc_state,
+                metadata=_encode_metadata(
+                    _update_site_metadata(
+                        site.metadata,
+                        inference_type=constants.INFERENCE_FITCH_PARSIMONY,
+                    )
+                ),
+            )
+            mut_map = {tskit.NULL: tskit.NULL}
+            if anc_state != inferred_anc_state:
+                # Need to add an extra mutation above the root to switch ancestral state
+                for root in tree.roots:
+                    mut_map[tskit.NULL] = tables.mutations.add_row(
+                        site=new_site_id,
+                        node=root,
+                        derived_state=inferred_anc_state,
+                        parent=tskit.NULL,
+                    )
+            for i, mutation in enumerate(mapped_mutations):
+                mut_map[i] = tables.mutations.add_row(
+                    site=new_site_id,
+                    node=mutation.node,
+                    derived_state=mutation.derived_state,
+                    parent=mut_map[mutation.parent],
+                )
+        pm.update()
+    pm.close()
+
+    tables.sort()
+    if check_consistency:
```

Comment:
We shouldn't do this here - this is  testing functionality, not user functionality.

Reply:
OK. I do think we want to allow the user to check that the samples (0..N) in the sample_data file correspond to the sample nodes (0..N) in the corresponding tree sequence, don't we? I can imagine a user calling this function on an unsimplified tree sequence and getting back a completely weird result, unless some checking is done. Similarly, are we guaranteeing that the sample nodes returned from tsinfer are in the same order as the samples in the sample_data file?

Reply:
If both have n samples, then that's a pretty good check, right?

Reply:
We shouldn't be relying on the samples in the ts being 0..N though.

Reply:
Aren't the samples sometimes at the end of the nodes array, rather than at the beginning? In general, I wasn't sure we should rely on users expecting the ids of the nodes in the sample data and the tree sequence being the same and in the same order (although we are OK when using the function internally, I suppose)

Reply:
The samples are just ``ts.samples()``, and they can be anywhere in the nodes array. It's part of the API contract that the user provides a tree sequence in which ts.samples() corresponds to the sample data file. The simplest check is that the sizes are equal. Let's just keep this simple.